perception/autoware_image_projection_based_fusion/src/camera_projection.cpp

```diff
@@ -0,0 +1,264 @@
+
+      const int grid_x = static_cast<int>(std::floor(qx / grid_w_size_));
+      const int grid_y = static_cast<int>(std::floor(qy / grid_h_size_));
+      const int index = (grid_y)*grid_x_num_ + grid_x;
+
+      // precompute projected point
+      cv::Point2d raw_image_point = camera_model_.unrectifyPoint(cv::Point2d(qx, qy));
+      projection_cache_[index] =
+        PixelPos{static_cast<float>(raw_image_point.x), static_cast<float>(raw_image_point.y)};
+    }
+  }
+}
+
+/**
+ * @brief Calculate a projection of 3D point to rectified image plane 2D point.
+ * @return Return a boolean indicating whether the projected point is on the image plane.
+ */
+bool CameraProjection::calcRectifiedImageProjectedPoint(
+  const cv::Point3d & point3d, Eigen::Vector2d & projected_point)
+{
+  const cv::Point2d rectified_image_point = camera_model_.project3dToPixel(point3d);
+
+  if (
+    rectified_image_point.x < 0.0 || rectified_image_point.x >= image_w_ ||
+    rectified_image_point.y < 0.0 || rectified_image_point.y >= image_h_) {
+    return false;
+  } else {
+    projected_point << rectified_image_point.x, rectified_image_point.y;
+    return true;
+  }
+}
+
+/**
+ * @brief Calculate a projection of 3D point to raw image plane 2D point.
+ * @return Return a boolean indicating whether the projected point is on the image plane.
+ */
+bool CameraProjection::calcRawImageProjectedPoint(
+  const cv::Point3d & point3d, Eigen::Vector2d & projected_point)
+{
+  const cv::Point2d rectified_image_point = camera_model_.project3dToPixel(point3d);
+  const cv::Point2d raw_image_point = camera_model_.unrectifyPoint(rectified_image_point);
+
+  if (
+    rectified_image_point.x < 0.0 || rectified_image_point.x >= image_w_ ||
+    rectified_image_point.y < 0.0 || rectified_image_point.y >= image_h_) {
+    return false;
+  } else {
+    projected_point << raw_image_point.x, raw_image_point.y;
+    return true;
+  }
+}
+
+/**
+ * @brief Calculate a projection of 3D point to raw image plane 2D point with approximation.
+ * @return Return a boolean indicating whether the projected point is on the image plane.
+ */
+bool CameraProjection::calcRawImageProjectedPointWithApproximation(
+  const cv::Point3d & point3d, Eigen::Vector2d & projected_point)
+{
+  const cv::Point2d rectified_image_point = camera_model_.project3dToPixel(point3d);
+
+  // round to a near grid center
+  const float qx =
+    std::round((rectified_image_point.x - half_grid_w_size_) * inv_grid_w_size_) * grid_w_size_ +
+    half_grid_w_size_;
+  if (qx < 0.0 || qx >= image_w_) {
+    return false;
+  }
+  const float qy =
+    std::round((rectified_image_point.y - half_grid_h_size_) * inv_grid_h_size_) * grid_h_size_ +
+    half_grid_h_size_;
+  if (qy < 0.0 || qy >= image_h_) {
+    return false;
+  }
+
+  const int grid_x = static_cast<int>(std::floor(qx / grid_w_size_));
+  const int grid_y = static_cast<int>(std::floor(qy / grid_h_size_));
```

Comment:
```suggestion
  // round to a near grid center
  const float qx =
    std::round((rectified_image_point.x - half_grid_w_size_) * inv_grid_w_size_)  + 0.5f;
  if (qx < 0.0 || qx >= image_w_ /grid_w_size_) {
    return false;
  }
  const float qy =
    std::round((rectified_image_point.y - half_grid_h_size_) * inv_grid_h_size_)  + 0.5f;
  if (qy < 0.0 || qy >= image_h_/grid_h_size_) {
    return false;
  }

  const int grid_x = static_cast<int>(std::floor(qx));
  const int grid_y = static_cast<int>(std::floor(qy));
```

may it can be more efficient
(need to be tested)

Reply:
```suggestion
  // round to a near grid center
  const int grid_x = std::floor(rectified_image_point.x * inv_grid_w_size_);
  const int grid_y = std::floor(rectified_image_point.y * inv_grid_h_size_);
  
  if  (grid_x < 0.0 || grid_x>= static_cast<int>(image_w_/grid_w_size_)) return false;
  if  (grid_y < 0.0 || grid_y>= static_cast<int>(image_h_/grid_h_size_)) return false;
```

more simple like this? is it the same?

Reply:
Thank you for the great suggestion!
I confirm the suggestions result in same values and behavior except for the (0, \*), (\*, 0) pixels.
(0, \*), (\*, 0) pixels will be get inside the grid which is better than the original one.

I fix some of the code and added in https://github.com/autowarefoundation/autoware.universe/pull/9635/commits/e8b5b2ffdcfcca886f51c9d58f26f623356a5932